File: stepik-union/src/main/java/org/stepik/core/courseFormat/stepHelpers/SortingQuizHelper.java
package org.stepik.core.courseFormat.stepHelpers;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Pair;
import org.jetbrains.annotations.NotNull;
import org.stepik.core.courseFormat.StepNode;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/**
 * @author meanmail
 */
public class SortingQuizHelper extends QuizHelper {
    private List<Pair<Integer, String>> ordering;

    public SortingQuizHelper(@NotNull Project project, @NotNull StepNode stepNode) {
        super(project, stepNode);
    }

    @SuppressWarnings("unused")
    @NotNull
    public List<Pair<Integer, String>> getOrdering() {
        initStepOptions();
        return ordering;
    }

    @Override
    protected void done() {
        List<Integer> replyOrdering = reply.getOrdering();
        List<String> values = getDataset().getOptions();

        int valuesCount = values.size();

        if (replyOrdering.size() != valuesCount) {
            replyOrdering = IntStream.range(0, valuesCount)
                    .boxed()
                    .collect(Collectors.toList());
        }
        ordering = replyOrdering.stream()
                .map(index -> Pair.create(index, index < valuesCount ? values.get(index) : ""))
                .collect(Collectors.toList());
    }

    @Override
    void fail() {
        ordering = new ArrayList<>();
    }

    @Override
    public boolean isAutoCreateAttempt() {
        return true;
    }
}

File: stepik-union/src/main/java/org/stepik/core/courseFormat/stepHelpers/SortingQuizHelper.kt
package org.stepik.core.courseFormat.stepHelpers

import com.intellij.openapi.project.Project
import com.intellij.openapi.util.Pair
import org.stepik.core.courseFormat.StepNode


class SortingQuizHelper(project: Project, stepNode: StepNode) : QuizHelper(project, stepNode) {
    private var ordering: List<Pair<Int, String>>? = null

    fun getOrdering(): List<Pair<Int, String>> {
        initStepOptions()
        return ordering!!
    }

    override fun done() {
        var replyOrdering = reply.ordering
        val values = dataset.options

        val valuesCount = values.size

        if (replyOrdering.size != valuesCount) {
            replyOrdering = (0..valuesCount).toList()
        }
        ordering = replyOrdering.map { index -> Pair.create(index, if (index < valuesCount) values[index] else "") }
    }

    override fun fail() {
        ordering = emptyList()
    }
}
